lib/reek/smells/nested_iterators.rb

```diff
@@ -79,31 +80,36 @@ def find_deepest_iterator(ctx)
       #   foo.each() do ... end
       # this will end up as:
       #
-      # "foo.each() do ... end" -> the iterator below
-      # "each()"                -> the "call" below
-      # "do ... end"            -> the "block" below
-      #
-      # @param parent [AST::Node] The parent iterator
+      # "foo.each() do ... end" -> one of the :block nodes
+      # "each()"                -> the node's "call"
+      # "do ... end"            -> the node's "block"
       #
       # @param exp [AST::Node]
       #   The given expression to analyze.
-      #   Will be nil on empty blocks so we'll return just the parent iterator
       #
       # @param depth [Integer]
       #
       # @return [Array<Iterator>]
       #
-      def scout(parent:, exp:, depth:)
-        return [Iterator.new(parent, depth)] unless exp
-        iterators = exp.find_nodes([:block])
-        return [Iterator.new(parent, depth)] if iterators.empty?
-        iterators.map do |iterator|
-          # 1st case: we recurse down the given block of the iterator. In this case
-          # we need to check if we should increment the depth.
-          # 2nd case: we recurse down the associated call of the iterator. In this case
-          # the depth stays the same.
-          scout(parent: iterator, exp: iterator.block, depth: increment_depth(iterator, depth)) +
-            scout(parent: iterator, exp: iterator.call, depth: depth)
+      def scout(exp:, depth:)
+        return [] unless exp
+        exp.find_nodes([:block]).flat_map do |node|
+          examine_block_node(node, depth)
+        end
+      end
+
+      def examine_block_node(iterator, depth)
+        new_depth = increment_depth(iterator, depth)
+        # 1st case: we recurse down the given block of the iterator. In this case
+        # we need to check if we should increment the depth.
+        block_result = scout(exp: iterator.block, depth: new_depth)
```

Comment:
On the other hand I don't want to block this improvement just because of that. But isn't there a way to not have circular calls and instead use recursion like before?